src/platform/Darwin/DnssdImpl.cpp

```diff
@@ -162,17 +164,45 @@ MdnsContexts MdnsContexts::sInstance;
 
 namespace {
 
+static void OnRegisterRecord(DNSServiceRef sdRef, DNSRecordRef recordRef, DNSServiceFlags flags, DNSServiceErrorType err,
+                             void * context)
+{
+    ChipLogDetail(Discovery, "Mdns: %s Registered Record err %d", __func__, err);
+    auto sdCtx = reinterpret_cast<RegisterContext *>(context);
+    sdCtx->Finalize(err);
```

Comment:
I think we would need a dedicated `RegisterRecordContext` here. My understanding of the current code is that we are reusing the `RegisterContext` and so we may fire the success/failure callback multiple times ortherwise.

Reply:
done